qubes-gui-agent-proto/src/lib.rs

```diff
@@ -0,0 +1,195 @@
+pub enum Error {
+    /// Invalid UTF-8
+    BadUTF8(core::str::Utf8Error),
+    /// Invalid key event type
+    BadKeypress {
+        /// The type provided by the GUI daemon
+        ty: u32,
+    },
+    /// Invalid button event status
+    BadButton {
+        /// The type provided by the GUI daemon
+        ty: u32,
+    },
+    /// Invalid focus event status
+    BadFocus {
+        /// The type provided by the GUI daemon
+        ty: u32,
+    },
+}
+
+/// An event that a GUI agent must handle
+#[non_exhaustive]
+pub enum DaemonToAgentEvent<'a> {
+    /// The pointer has moved
+    Motion {
+        /// The contents of the event
+        event: qubes_gui::Motion,
+    },
+    /// The pointer has entered or left a window.
+    Crossing {
+        /// The contents of the event
+        event: qubes_gui::Crossing,
+    },
+    /// The user wishes to close a window
+    Close,
+    /// A key has been pressed or released
+    Keypress {
+        /// The contents of the event
+        event: qubes_gui::Keypress,
+    },
+    /// A button has been pressed or released
+    Button {
+        /// The contents of the event
+        event: qubes_gui::Button,
+    },
+    /// The GUI daemon has requested the contents of the clipboard.  The agent
+    /// is expected to send a [`MSG_CLIPBOARD_DATA`] message with the
+    /// requested data.
+    Copy,
+    /// Set the contents of the clipboard.
+    Paste {
+        /// The pasted data, which is not trusted
+        untrusted_data: &'a str,
+    },
+    /// The keymap has changed.
+    Keymap {
+        /// The new keymap
+        new_keymap: qubes_gui::KeymapNotify,
+    },
+    /// The agent must redraw a portion of the display
+    Redraw {
+        /// The portion of the window to redraw
+        portion_to_redraw: qubes_gui::MapInfo,
+    },
+    /// A window has been moved and/or resized.
+    Configure {
+        /// The contents of the event
+        new_size_and_position: qubes_gui::Configure,
+    },
+    /// A window has gained or lost focus
+    Focus {
+        /// The contents of the event
+        event: qubes_gui::Focus,
+    },
+    /// Window manager flags have changed
+    WindowFlags {
+        /// The contents of the event
+        flags: qubes_gui::WindowFlags,
+    },
+    /// GUI daemon confirms window destruciton; window ID may be reused
+    Destroy,
+}
+
+impl<'a> DaemonToAgentEvent<'a> {
+    /// Parse a Qubes OS GUI message from the GUI daemon
+    ///
+    /// # Panics
+    ///
+    /// Will panic if the length of the message does not match the length in the
+    /// header.  May (or may not!) panic if the length of the message is not valid.
+    /// The caller is expected to have validated this earlier, using
+    /// [`qubes_gui::msg_length_limits`].
+    ///
+    /// # Errors
+    ///
+    /// Fails if the given GUI message cannot be parsed.
+    pub fn parse(header: qubes_gui::Header, body: &'a [u8]) -> Result<Option<(u32, Self)>, Error> {
```

Comment:
I'd expect an explanation of why we're using both `Result` and `Option` here, and in particular the semantics of an `Ok(None)` return value.